src/goto-instrument/cover_boundary_values.cpp

```diff
@@ -0,0 +1,314 @@
+/*******************************************************************\
+
+Module: Coverage Instrumentation for Boundary Values
+
+Author: Youcheng Sun
+
+Date: Oct 2016
+
+\*******************************************************************/
+
+#include <algorithm>
+#include <iterator>
+
+#include <util/prefix.h>
+#include <util/i2string.h>
+#include <util/expr_util.h>
+
+#include "cover.h"
+
+/*******************************************************************\
+
+Function: non_ordered_predicate_expansion
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: expand a non-ordered predicate: <=, !=, >=
+
+\*******************************************************************/
+
+std::set<exprt> non_ordered_predicate_expansion(const exprt &src)
+{
+  std::set<exprt> result;
+  // the expansion of "<=" is "<" and "=="
+  if(src.id()==ID_le)
+  {
+    exprt e1(ID_lt);
+    e1.type().id(src.type().id());
+    e1.operands().push_back(src.op0());
+    e1.operands().push_back(src.op1());
```

Comment:
Wouldn't it be simpler to copy the entire expression and replace the id()? That is, do exprt e1(src); e1.id(ID_lt);